src/pg_query_summary.c

```diff
@@ -0,0 +1,775 @@
+
+			break;
+		}
+
+		// Things which we are only handling because raw_expression_tree_walker()
+		// doesn't handle them.
+
+		case T_CheckPointStmt:
+		case T_CreateSchemaStmt:
+		case T_GrantRoleStmt:
+		case T_TransactionStmt:
+		case T_VariableShowStmt:
+		case T_VariableSetStmt:
+			return false;
+
+		default:
+			break;
+	}
+
+	return raw_expression_tree_walker(node, pg_query_summary_walk_impl, (void*) state);
+}
+
+static char* range_var_parts_to_name(char* schemaname, char* relname) {
+	StringInfoData buf;
+	initStringInfo(&buf);
+	if (schemaname != NULL) {
+		appendStringInfoString(&buf, quote_identifier(schemaname));
+		appendStringInfoChar(&buf, '.');
+	}
+	appendStringInfoString(&buf, quote_identifier(relname));
+	return buf.data;
+}
+
+// Walk through all collected cte_names, and compares them to relname.
+//
+// Returns true if any are an exact match, false otherwise.
+static bool cte_names_include(WalkState* state, char* relname) {
+	ListCell* lc = NULL;
+	foreach(lc, state->cte_names) {
+		char *ctename = lfirst(lc);
+		if (strcmp(ctename, relname) == 0)
+			return true;
+	}
+	return false;
+}
+
+static void handle_from_clause_item(Node* node, ContextType context, WalkState* state) {
+	if (node == NULL)
+		return;
+
+	switch (nodeTag(node)) {
+		case T_JoinExpr: {
+			JoinExpr* je = castNode(JoinExpr, node);
+
+			handle_from_clause_item((Node*) je->larg, context, state);
+			handle_from_clause_item((Node*) je->rarg, context, state);
+
+			handle_from_clause_item(je->quals, context, state);
+
+			break;
+		}
+		case T_RangeVar: {
+			RangeVar* rv = castNode(RangeVar, node);
+			add_table(state, context, castNode(RangeVar, node));
+
+			if (rv->alias) {
+				SummaryAlias* alias = palloc(sizeof(SummaryAlias));
+				alias->key = rv->alias->aliasname;
+				alias->value = range_var_parts_to_name(rv->schemaname, rv->relname);
+				state->aliases = lappend(state->aliases, alias);
+			}
+
+			break;
+		}
+		default:
+			break;
+	}
+}
+
+// Walk the parse tree, storing the results in `summary`.
+//
+// Returns NULL on success.
+PgQueryError* pg_query_summary_walk(Summary* summary, Node* tree) {
+	WalkState state = {0};
+
+	// NOTE regarding cte_names and from_clause_items:
+	// 1. We process cte_names as we iterate through the tree.
+	// 2. We only add items to tables if they are *not* in cte_names.
+	// 3. CTEs can be referenced *after* table names.
+	// 4. Because of this, we have to store from_clause_items and process it after the initial tree walk.
+	//
+	// If we try to do it entirely on the initial pass, we wind up with CTEs being
+	// included in state->tables, and thus in summary->tables.
+
+	pg_query_summary_walk_impl(tree, &state);
+
+	ListCell* lc = NULL;
+	foreach(lc, state.from_clause_items) {
+		NodeWithContext* lc_node_with_context = lfirst(lc);
+
+		Node* lc_node = lc_node_with_context->node;
+
+		handle_from_clause_item(lc_node, lc_node_with_context->context, &state);
```

Comment:
Per the other notes, if we special case this to just be range vars, it may also be nice to simplify the call here to pass the `RangeVarWithContext` directly and have the (renamed) `handle_range_var_with_context` take that.